### ISSUE TITLE
[BUG] Different behavior in 0.1.8 and 0.2.1

### ISSUE THREAD

**Issue by kuilef** [BrikerMan/Kashgari#55]
## Environment

* Colab.research.google.com
* Kashgari 1.1.8 / 0.2.1

## Issue Description
Different behavior in 0.1.8 and 0.2.1
In Kashgari 0.1.8 BLSTModel converge in training process and I see val_acc: 0.98 and train_acc: 0.9594
In Kashgari 0.2.1 BLSTModel is overfitting and I see val_acc ~0.5 and train_acc ~0.96
There is no difference in my code, only different versions of library.

### Reproduce
code:

```
from sklearn.model_selection import train_test_split
import pandas as pd
import nltk
from kashgari.tasks.classification import BLSTMModel

# get and process data
!wget https://www.dropbox.com/s/265kphxkijj1134/fontanka.zip

df1 = pd.read_csv('fontanka.zip')
df1.fillna(' ', inplace = True)
nltk.download('punkt')

# split on train/test
X_train, X_test, y_train, y_test = train_test_split(df1.full_text[:3570].values, df1.textrubric[:3570].values, test_size=0.2, random_state=42)
X_train = [nltk.word_tokenize(sentence) for sentence in X_train]
X_test  = [nltk.word_tokenize(sentence) for sentence in X_test]
y_train = y_train.tolist()
y_test  = y_test.tolist()

# train model
model = BLSTMModel()
model.fit(X_train, y_train, x_validate=X_test, y_validate=y_test, epochs = 10)
```

code in colab:
https://colab.research.google.com/drive/1yTBMeiBl2y7-Yw0DS_vTn2A4y_Vj3N-8

### Result
Last epoch:

**Kashgari 0.1.8**
> Epoch 10/10
> 55/55 [==============================] - 90s 2s/step - loss: 0.1378 - acc: 0.9615 - val_loss: 0.0921 - val_acc: 0.9769

**Kashgari 0.2.1**
> Epoch 10/10
> 44/44 [==============================] - 76s 2s/step - loss: 0.0990 - acc: 0.9751 - val_loss: 2.3739 - val_acc: 0.5323

### Other Comment
In 0.2.1 all models now in different file and lr hyperparameter is given explicitly (1e-3)
In 0.1.8 lr hyperparameter was omitted, I suppose that it used keras default, which is the same (1e-3)

Also in 0.1.8 you had (dense size = +1 classes on classifier)
https://github.com/BrikerMan/Kashgari/issues/21 and ommited it in 0.2.1. I don't see how this could affect training process.

I couldn't find more differences between versions, could you help with this - why models began to overfit in new version of library?

**Comment by alexwwang** [comment]
Have you ever tried to train your model twice in a row with exactly the
same parameters configuration?

In my experience, the optimizer `adam`, as well as `sgd` would not be
exactly the same on convergence track in each training of a given model,
for the randomness of the algorithm itself.

And yes, the author eliminated the place holder class in classification
model and this may also cause the change to optimization, so as to the
result.

The explicitly assigned parameters are the same to the keras defaults, here
we just provide a convenient and consistent interface for users to
initialize new models with different hyperparameters.

**Comment by kuilef** [comment]
Yes, I trained model several times with unchanged parameters, the result is about the same, in 0.2.1 network always overfitting. Also I tried BLSTMModel with embedding (BERT) - same results.

**Comment by BrikerMan** [comment]
I have reproduced the problem. 0.2.1 does overfit in my dataset too. I have compared 0.1.8 with 0.2.1, we have changed BLSTMModel's activation function from `sigmoid` to `softmax`.
Please try this with 0.2.1.

```
hyper_parameters = {
    'activation_layer': {
        'activation': 'sigmoid'
    }
}
model1 = BLSTMModel(hyper_parameters=hyper_parameters)

model2 = BLSTMModel()
```

**Comment by alexwwang** [comment]
This is interesting. I will try it too.

**Comment by kuilef** [comment]
Thanks, @BrikerMan I didn't touch anything else and tried this
```
hyper_parameters = {
    'activation_layer': {
        'activation': 'sigmoid'
    }
}
model = BLSTMModel(hyper_parameters=hyper_parameters)
```

```
{'architect_name': 'BLSTMModel',
 'embedding': {'embedding_size': 100,
  'embedding_type': 'base',
  'kwargs': {},
  'model_path': '',
  'name': 'custom',
  'sequence_length': 0},
 'hyper_parameters': {'activation_layer': {'activation': 'sigmoid'},
  'compile_params': {'loss': 'categorical_crossentropy',
   'metrics': ['accuracy']},
  'lstm_layer': {'return_sequences': False, 'units': 256},
  'optimizer': {'module': 'keras.optimizers',
   'name': 'adam',
   'params': {'decay': 0.0, 'lr': 0.001}}},
 'model_info': {'multi_label': False},
 'task': 'classification'}

```
Strange, but it doesn't help, model overfitting again :(

**Comment by BrikerMan** [comment]
Ok, Maybe we should compare model structure.  Please try `model.model.to_json()` for both model after fit one epoch for each model.

**Comment by kuilef** [comment]
I ran model.model.to_json() for both models after fitting one epoch for each model.
JSON data for both models is the same except **batch_input_shape** and **last Dense layer**.

in kashgari 0.1.8 there is 367 neurons in inputlayer and 17 in outputlayer.
in kashgari 0.2.1 there is 374 neurons in inputlayer and 16 in outputlayer.
I have 16 classes in dataset.

**Comment by alexwwang** [comment]
Maybe we should put the dummy class back.
@BrikerMan

**Comment by BrikerMan** [comment]
Let's try one more thing.

Please install the development version by
```bash
pip install git+https://github.com/BrikerMan/Kashgari.git@develop
```

Then 

```python
t_model = CNNModel()

# Make the input layer has equal neurons
t_model.embedding._sequence_length = 367

# remember to pass validate data if u have validate data
t_model.build_token2id_label2id_dict(train_x, train_y)

# add dummy class manually
idx2label = t_model._idx2label
idx2label[len(idx2label)] = idx2label[0]
idx2label[0] = '[PAD]'
label2idx = dict([(val, key) for (key, val) in idx2label.items()])
t_model.label2idx = label2idx

# Then train 
```

**Comment by kuilef** [comment]
I tried your solution and added sigmoid activation too.
```
hyper_parameters = {
    'activation_layer': {
        'activation': 'sigmoid'
    }
}

t_model = BLSTMModel(hyper_parameters=hyper_parameters)

# Make the input layer has equal neurons
t_model.embedding._sequence_length = 367

# remember to pass validate data if u have validate data
t_model.build_token2id_label2id_dict(X_train, y_train, x_validate=X_test, y_validate=y_test)

# add dummy class manually
idx2label = t_model._idx2label
idx2label[len(idx2label)] = idx2label[0]
idx2label[0] = '[PAD]'
label2idx = dict([(val, key) for (key, val) in idx2label.items()])
t_model.label2idx = label2idx

# Then train 
t_model.fit(X_train, y_train, x_validate=X_test, y_validate=y_test, epochs = 10)
```

I ran two models (0.1.8 and 0.2.1) and compared their json model structures - identical!
But again in 0.1.8 is fine
```
Epoch 10/10
55/55 [==============================] - 75s 1s/step - loss: 0.1857 - acc: 0.9536 - val_loss: 0.1328 - val_acc: 0.9738
```

in 0.2.1 overfitting
```
Epoch 11/11
44/44 [==============================] - 74s 2s/step - loss: 0.1831 - acc: 0.9523 - val_loss: 1.9028 - val_acc: 0.5569
```

Here is the link for code in colab with all models processed and structure in json also.
I'm splitting data with train_test_split and constant random_state, so the data is the same.
I don't understand :(
https://colab.research.google.com/drive/1oC66NTZG52NnPuVjmz8dFpVoIJ3woF9A

**Comment by alexwwang** [comment]
Have you ever tried to train this model locally with your own CPU?

**Comment by kuilef** [comment]
@alexwwang thanks for your idea, I tried it with tensorflow 1.3.1
I tried both locally on CPU and on my small GPU (GeForce 1030, tensorflow-gpu)

0.2.1 with sigmoid activation function in hyper_parameters is overfitting.
0.2.1 development version with sigmoid activation function in hyper_parameters and with manual dummy class is overfitting.

**Comment by alexwwang** [comment]
Have you ever tried to change the readable alterations of related part of codes in 0.2.1 develop branch back to 0.1.8 to see if the problem stays?

**Comment by kuilef** [comment]
Sorry, @BrikerMan but problem with overfitting persist in 0.2.4
![image](https://user-images.githubusercontent.com/19805977/59692692-2cb7d580-91ee-11e9-9783-347033e4faf9.png)

**Comment by BrikerMan** [comment]
Please try the tf.keras version... I have fixed an overfitting issue on
that branch.

Ilya Kuznetsov <notifications@github.com>于2019年6月18日 周二22:26写道：

**Comment by kuilef** [comment]
I'm starting to think that maybe something wrong was in 0.1.8 when val_acc was 0.98 in BLSTModel()
I tested on tf.keras version -- all the same as in 0.2.4, 0.2.1 - I tested version with embeddings and without. 
I tried to mimic the same configuration with
```
hyper_parameters = {
    'layer_bi_lstm': {'return_sequences': False, 'units': 256},
    'layer_dense': {'activation': 'sigmoid'}
}
model = BLSTMModel(hyper_parameters=hyper_parameters)
```

but the model does overfit.

**Comment by BrikerMan** [comment]
Really sorry to hear that, I have tested 0.2.4 and tf.keras version on two datasets, all works just fine. Maybe it is a bug in 0.1.8.
